include/fuzzuf/algorithms/nautilus/grammartec/tree.hpp

```diff
@@ -0,0 +1,174 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2022 Ricerca Security
+ * 
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+/**
+ * @file tree.hpp
+ * @brief Tree for context-free grammar
+ * @author Ricerca Security <fuzzuf-dev@ricsec.co.jp>
+ */
+#ifndef FUZZUF_INCLUDE_ALGORITHMS_NAUTILUS_GRAMMARTEC_TREE_HPP
+#define FUZZUF_INCLUDE_ALGORITHMS_NAUTILUS_GRAMMARTEC_TREE_HPP
+
+#include <optional>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <variant>
+#include <vector>
+#include "fuzzuf/algorithms/nautilus/grammartec/context.hpp"
+#include "fuzzuf/algorithms/nautilus/grammartec/newtypes.hpp"
+#include "fuzzuf/algorithms/nautilus/grammartec/rule.hpp"
+
+
+namespace fuzzuf::algorithm::nautilus::grammartec {
+
+class RecursionInfo;
+class Tree;
+class TreeLike;
+class TreeMutation;
+class Unparser;
+class UnparserStep;
+
+
+class TreeLike {
+public:
```

Comment:
base class should have a virtual destructor even when it's empty.

Reply:
Fixed in [29474e7](https://github.com/fuzzuf/fuzzuf/pull/23/commits/29474e7f48c06242224c5734dd0d3c70030817dc)